File: backend/src/main/java/com/horain/agent/AgentFeedbackService.java
package com.horain.agent;

import com.horain.model.AgentFeedback;
import com.horain.model.AgentTurn;
import com.horain.repository.AgentFeedbackRepository;
import com.horain.repository.AgentTurnRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

/**
 * Saves or updates user feedback (thumb up/down) on agent turns.
 */
@Service
public class AgentFeedbackService {

    private final AgentFeedbackRepository feedbackRepository;
    private final AgentTurnRepository turnRepository;

    public AgentFeedbackService(AgentFeedbackRepository feedbackRepository, AgentTurnRepository turnRepository) {
        this.feedbackRepository = feedbackRepository;
        this.turnRepository = turnRepository;
    }

    /**
     * Creates or updates feedback for the given turn. Validates that the turn exists and rating is "up" or "down".
     */
    @Transactional
    public AgentFeedback saveFeedback(UUID turnId, String rating, String reasonCode, String comment) {
        if (!turnRepository.existsById(turnId)) {
            throw new IllegalArgumentException("Turn not found: " + turnId);
        }
        String normalizedRating = rating != null ? rating.strip().toLowerCase() : "";
        if (!"up".equals(normalizedRating) && !"down".equals(normalizedRating)) {
            throw new IllegalArgumentException("Rating must be 'up' or 'down'");
        }
        AgentFeedback feedback = feedbackRepository.findByTurnId(turnId).orElse(null);
        if (feedback == null) {
            feedback = new AgentFeedback();
            feedback.setId(UUID.randomUUID());
            feedback.setTurnId(turnId);
        }
        feedback.setRating(normalizedRating);
        feedback.setReasonCode(reasonCode != null && !reasonCode.isBlank() ? reasonCode : null);
        feedback.setComment(comment != null && !comment.isBlank() ? comment : null);
        return feedbackRepository.save(feedback);
    }
}

File: backend/src/main/kotlin/com/horain/agent/AgentFeedbackService.kt
package com.horain.agent

import com.horain.model.AgentFeedback
import com.horain.repository.AgentFeedbackRepository
import com.horain.repository.AgentTurnRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

/**
 * Saves or updates user feedback (thumb up/down) on agent turns.
 */
@Service
class AgentFeedbackService(
    private val feedbackRepository: AgentFeedbackRepository,
    private val turnRepository: AgentTurnRepository
) {

    /**
     * Creates or updates feedback for the given turn. Validates that the turn exists and rating is "up" or "down".
     */
    @Transactional
    fun saveFeedback(turnId: UUID, rating: String?, reasonCode: String?, comment: String?): AgentFeedback {
        if (!turnRepository.existsById(turnId)) {
            throw IllegalArgumentException("Turn not found: $turnId")
        }
        val normalizedRating = rating?.trim()?.lowercase() ?: ""
        if (normalizedRating != "up" && normalizedRating != "down") {
            throw IllegalArgumentException("Rating must be 'up' or 'down'")
        }
        var feedback = feedbackRepository.findByTurnId(turnId).orElse(null)
        if (feedback == null) {
            feedback = AgentFeedback()
            feedback.id = UUID.randomUUID()
            feedback.turnId = turnId
        }
        feedback.rating = normalizedRating
        feedback.reasonCode = if (!reasonCode.isNullOrBlank()) reasonCode else null
        feedback.comment = if (!comment.isNullOrBlank()) comment else null
        return feedbackRepository.save(feedback)
    }
}
